internal/cloud/update_uploader.go

```diff
@@ -0,0 +1,128 @@
+package cloud
+
+import (
+	"context"
+	"time"
+
+	"github.com/golang/protobuf/ptypes"
+	"github.com/golang/protobuf/ptypes/duration"
+	"github.com/golang/protobuf/ptypes/timestamp"
+
+	"github.com/windmilleng/tilt/internal/feature"
+	"github.com/windmilleng/tilt/internal/store"
+)
+
+type UpdateUploader struct {
+	client HttpClient
+	addr   Address
+
+	lastCompletedBuildCount int
+	lastFinishTime          time.Time
+}
+
+func NewUpdateUploader(client HttpClient, addr Address) *UpdateUploader {
+	return &UpdateUploader{
+		client: client,
+		addr:   addr,
+	}
+}
+
+// TODO(nick): Generate these with protobufs.
+type updateServiceSpec struct {
+	Name string `json:"name"`
+}
+
+type update struct {
+	Service      updateServiceSpec    `json:"service"`
+	StartTime    *timestamp.Timestamp `json:"start_time"`
+	Duration     *duration.Duration   `json:"duration"`
+	IsLiveUpdate bool                 `json:"is_live_update"`
+
+	// 0 = SUCCESS, 1 = FAILURE
+	Result            int    `json:"result"`
+	ResultDescription string `json:"result_description"`
+
+	// TODO(nick): auto-create Snapshot IDs?
+}
+
+// Check the engine state to see if we have any updates.
+func (u *UpdateUploader) makeUpdates(st store.RStore) []update {
+	state := st.RLockState()
+	defer st.RUnlockState()
+
+	// Check if this feature is disabled
+	if !state.Features[feature.UpdateHistory] {
+		return nil
+	}
+
+	// Do a quick check to see if any builds have completed since we last checked.
+	if state.CompletedBuildCount == 0 || state.CompletedBuildCount <= u.lastCompletedBuildCount {
+		return nil
+	}
+
+	// OK, we know we have work to do!
+
+	highWaterMark := u.lastFinishTime
```

Comment:
is there anything protecting this function from getting invoked in parallel by multiple calls to `OnChange` and processing the same set of builds?

similarly, do we need synchronization on accessing `u`'s fields?

Reply:
from:
https://github.com/windmilleng/tilt/blob/master/internal/store/subscriber.go#L11

```
// Subscribers do not need to be thread-safe. The Store will only
// call OnChange for a given subscriber when the last call completes.
```

Reply:
Whoops, apparently my approach to subscriber concurrency has been off. I've been spinning blocking work off to a goroutine and then dealing with locking, unaware the whole thing is already in a thread-safe goroutine.